File: app/src/main/java/jp/kentan/student_portal_plus/ui/LectureInformationActivity.java
package jp.kentan.student_portal_plus.ui;

import android.app.Activity;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.animation.AnticipateOvershootInterpolator;
import android.widget.TextView;
import android.widget.Toast;

import jp.kentan.student_portal_plus.R;
import jp.kentan.student_portal_plus.data.PortalDataProvider;
import jp.kentan.student_portal_plus.data.component.LectureInformation;
import jp.kentan.student_portal_plus.notification.NotificationController;
import jp.kentan.student_portal_plus.util.LinkTransformationMethod;
import jp.kentan.student_portal_plus.util.StringUtils;

public class LectureInformationActivity extends AppCompatActivity {

    private LectureInformation mInfo;

    private boolean isNotificationMode;
    private boolean isMyClass = false;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_lecture_information);
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null){
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }


        /*
        Intent
         */
        Intent intent = getIntent();

        isNotificationMode = intent.getBooleanExtra("notification", false);
        try {
            if(isNotificationMode){
                new NotificationController(this).cancel(intent.getIntExtra("notify_id", -9));

                mInfo = PortalDataProvider.getLectureInfoByHash(this, intent.getStringExtra("hash"));
            }else{
                mInfo = PortalDataProvider.getLectureInfoById(intent.getIntExtra("id", 1));
            }
        } catch (Exception e) {
            failedLoad();
            return;
        }

        mInfo.setRead(true);
        isMyClass = mInfo.isMyClass();

        /*
        UI Initialize
         */
        final FloatingActionButton fab = (FloatingActionButton) findViewById(R.id.fab);

        final int status = mInfo.getMyClassStatus();
        if (status == LectureInformation.RESISTED_BY_PORTAL) {
            //受講解除できない場合はFAB隠す
            fab.hide();
        } else {
            if(status == LectureInformation.RESISTED_BY_SIMILAR){
                isMyClass = false;
            }

            final Activity activity = this;
            fab.setRotation((isMyClass) ? 135.0f : 0.0f);
            fab.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(final View view) {
                    if(isMyClass){
                        final AlertDialog.Builder builder = new AlertDialog.Builder(activity);
                        builder.setTitle("確認");
                        builder.setMessage(StringUtils.fromHtml(getString(R.string.msg_remove_my_class_info).replaceFirst("\\{subject\\}", mInfo.getSubject())));
                        builder.setPositiveButton(getString(R.string.yes), new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int which) {
                                if(mInfo.setMyClass(isMyClass=false)){
                                    Snackbar.make(view, "受講科目から除外しました", Snackbar.LENGTH_SHORT).setAction("Action", null).show();

                                    fab.animate().rotation(0.0f).setInterpolator(new AnticipateOvershootInterpolator()).withLayer().setDuration(800).start();
                                } else {
                                    final Snackbar snackbar = Snackbar.make(view, getString(R.string.err_msg_failed_to_remove_from_my_class), Snackbar.LENGTH_INDEFINITE);
                                    snackbar.setAction("閉じる", new View.OnClickListener() {
                                        @Override
                                        public void onClick(View v) {
                                            snackbar.dismiss();
                                        }
                                    }).show();
                                }
                            }
                        });
                        builder.setNegativeButton(getString(R.string.no), null);
                        builder.show();
                    }else{
                        mInfo.setMyClass(isMyClass=true);

                        Snackbar.make(view, "受講科目に追加しました", Snackbar.LENGTH_SHORT).setAction("Action", null).show();

                        fab.animate().rotation(135.0f).setInterpolator(new AnticipateOvershootInterpolator()).withLayer().setDuration(800).start();
                    }
                }
            });
        }


        setTitle(mInfo.getSubject());

        final TextView detail    = (TextView)findViewById(R.id.detail);

        String faculty = mInfo.getFaculty();
        String semester = mInfo.getSemester();

        faculty = (faculty.equals("-")) ? "" : faculty+" ";
        semester = (semester.equals("-")) ? "" : semester+" ";

        ((TextView)findViewById(R.id.subtitle)                   ).setText(mInfo.getInstructor());
        ((TextView)findViewById(R.id.subject)                    ).setText(mInfo.getSubject());
        ((TextView)findViewById(R.id.instructor)                 ).setText(mInfo.getInstructor());
        ((TextView)findViewById(R.id.semester_and_day_and_period)).setText(faculty + " " + semester + "  " + mInfo.getDayAndPeriod());
        ((TextView)findViewById(R.id.type)                       ).setText(mInfo.getType());
        ((TextView)findViewById(R.id.date)                     ).setText(mInfo.getDate());
        detail.setText(StringUtils.fromHtml(mInfo.getDetail()));

        detail.setTransformationMethod(new LinkTransformationMethod(this));
    }

    @Override
    public void onBackPressed() {
        if (isNotificationMode) {
            Intent home = new Intent(getApplicationContext(), HomeActivity.class);
            startActivity(home);
        }

        this.finish();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_share, menu);

        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_share:
                Intent intent = new Intent(Intent.ACTION_SEND);
                intent.setType("text/plain");
                intent.putExtra(Intent.EXTRA_SUBJECT, mInfo.getSubject());
                intent.putExtra(Intent.EXTRA_TEXT, createShareText());
                startActivity(intent);
                break;
            default:
                if (isNotificationMode) {
                    Intent home = new Intent(getApplicationContext(), HomeActivity.class);
                    startActivity(home);
                } else {
                    finish();
                }
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    private String createShareText() {
        return "授業科目:" + mInfo.getSubject() +
                "\n担当教員:" +
                mInfo.getInstructor() +
                "\n曜日時限:" +
                mInfo.getDayAndPeriod() +
                "\n分類:" +
                mInfo.getType() +
                "\n連絡事項:" +
                StringUtils.removeHtmlTag(mInfo.getDetail()) +
                "\n" +
                mInfo.getDate();
    }

    private void failedLoad() {
        Toast.makeText(this, getString(R.string.err_msg_failed_to_load), Toast.LENGTH_LONG).show();
        finish();
    }
}

File: app/src/main/java/jp/kentan/studentportalplus/ui/LectureInformationActivity.kt
package jp.kentan.studentportalplus.ui

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProvider
import android.os.Bundle
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.view.Menu
import android.view.MenuItem
import android.view.animation.AnticipateOvershootInterpolator
import dagger.android.AndroidInjection
import jp.kentan.studentportalplus.R
import jp.kentan.studentportalplus.data.component.LectureAttendType
import jp.kentan.studentportalplus.data.model.LectureInformation
import jp.kentan.studentportalplus.ui.viewmodel.LectureInformationViewModel
import jp.kentan.studentportalplus.ui.viewmodel.ViewModelFactory
import jp.kentan.studentportalplus.util.CustomTransformationMethod
import jp.kentan.studentportalplus.util.htmlToSpanned
import jp.kentan.studentportalplus.util.indefiniteSnackbar
import jp.kentan.studentportalplus.util.toShortString
import kotlinx.android.synthetic.main.activity_lecture_information.*
import kotlinx.android.synthetic.main.content_lecture_information.*
import org.jetbrains.anko.design.snackbar
import org.jetbrains.anko.longToast
import javax.inject.Inject


class LectureInformationActivity : AppCompatActivity() {

    private companion object {
        const val ROTATION_FROM =   0f
        const val ROTATION_TO   = 135f
        const val DURATION      = 800L
        val INTERPOLATOR = AnticipateOvershootInterpolator()
    }

    @Inject
    lateinit var viewModelFactory: ViewModelFactory

    private val viewModel by lazy {
        ViewModelProvider(this, viewModelFactory).get(LectureInformationViewModel::class.java)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_lecture_information)
        AndroidInjection.inject(this)

        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        viewModel.lectureInfo.observe(this, Observer<LectureInformation> { data ->
            if (data == null) {
                longToast(getString(R.string.error_not_found, getString(R.string.name_lecture_info)))
                finish()
                return@Observer
            }

            toolbar_layout.title = data.subject
            initView(data)

            if (data.attend == LectureAttendType.PORTAL) {
                fab.hide()
            } else {
                fab.show()
            }

            fab.setOnClickListener {
                val type = viewModel.getCurrentAttendType() ?: return@setOnClickListener

                if (type == LectureAttendType.USER) {
                    showConfirmationDialog(data.subject)
                } else {
                    viewModel.onClickAttendToUser { success ->
                        if (success) {
                            fab.animate()
                                    .rotation(ROTATION_TO)
                                    .setDuration(DURATION)
                                    .setInterpolator(INTERPOLATOR)
                                    .withLayer()
                                    .start()

                            snackbar(it, R.string.msg_register_class)
                        } else {
                            indefiniteSnackbar(it,
                                    getString(R.string.error_add, getString(R.string.name_attend_lecture)),
                                    getString(R.string.action_close))
                        }
                    }
                }
            }
        })

        viewModel.lectureInfoId.value = intent.getLongExtra("id", 0)
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.share, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        when (item?.itemId) {
            R.id.action_share -> viewModel.onClickShare(this)
            android.R.id.home -> finish()
        }

        return super.onOptionsItemSelected(item)
    }

    private fun initView(data: LectureInformation) {
        subject.text = data.subject
        instructor.text = data.instructor
        semester_week_period.text =
                getString(R.string.text_semester_week_period,
                        data.grade,
                        data.semester.formatSemester(),
                        data.week.formatWeek(),
                        data.period.formatPeriod())
        category.text = data.category
        detail.text = data.detailHtml.htmlToSpanned()
        detail.transformationMethod = CustomTransformationMethod(this)
        date.text = getString(R.string.text_created_date_lecture_info, data.createdDate.toShortString())

        if (data.createdDate != data.updatedDate) {
            date.append(getString(R.string.text_updated_date_lecture_info, data.updatedDate.toShortString()))
        }

        fab.rotation = if (data.attend == LectureAttendType.USER) ROTATION_TO else ROTATION_FROM
    }

    private fun showConfirmationDialog(subject: String) {
        val builder = AlertDialog.Builder(this)
        builder.setTitle(R.string.title_confirmation)
        builder.setMessage(getString(R.string.text_unregister_confirm, subject).htmlToSpanned())
        builder.setPositiveButton(R.string.action_yes) { _, _ ->
            viewModel.onClickAttendToNot { success ->
                if (success) {
                    snackbar(layout, R.string.msg_unregister_class)

                    fab.animate()
                            .rotation(ROTATION_FROM)
                            .setDuration(DURATION)
                            .setInterpolator(AnticipateOvershootInterpolator())
                            .withLayer()
                            .start()
                } else {
                    indefiniteSnackbar(layout,
                            getString(R.string.error_remove, getString(R.string.name_attend_lecture)),
                            getString(R.string.action_close))
                }
            }
        }
        builder.setNegativeButton(R.string.action_no, null)
        builder.show()
    }

    private fun String.formatSemester() = if (this == "前" || this == "後") this + "学期" else this.hyphenToWhitespace()

    private fun String.formatWeek() = this.replace("曜日", "曜")

    private fun String.formatPeriod() = if (this != "-") this + "限" else ""

    private fun String.hyphenToWhitespace() = if (this == "-") " " else this
}
